misc-utils/lsfd-sock-xinfo.c

```diff
@@ -404,43 +405,68 @@ struct inet_xinfo {
 	struct in_addr remote_addr;
 };
 
-static bool inet_fill_column(struct proc *proc __attribute__((__unused__)),
-			     struct inet_xinfo *inet,
-			     struct sock *sock __attribute__((__unused__)),
-			     struct libscols_line *ln __attribute__((__unused__)),
-			     int column_id,
-			     size_t column_index __attribute__((__unused__)),
-			     char **str)
+static uint32_t kernel32_to_cpu(enum sysfs_byteorder byteorder, uint32_t v)
 {
-	struct in_addr *nptr = NULL;
-	char s[INET_ADDRSTRLEN];
+	if (byteorder == SYSFS_BYTEORDER_LITTLE)
+		return le32_to_cpu(v);
+	else
+		return be32_to_cpu(v);
+}
 
-	switch(column_id) {
-	case COL_INET_LADDR:
-		nptr = &inet->local_addr;
-		break;
-	case COL_INET_RADDR:
-		nptr = &inet->remote_addr;
-		break;
-	default:
-		return false;
-	}
+/*
+ * L3 abstract-layer stacked on IP and IP6.
+ */
+struct l3_xinfo {
+	struct inet_xinfo inet;
+	unsigned int st;
+};
 
-	if (nptr && inet_ntop(AF_INET, nptr, s, sizeof(s))) {
-		*str = strdup(s);
-		return true;
-	}
-	return false;
-}
+struct l3_xinfo_class {
+	struct sock_xinfo_class sock;
+	struct sock_xinfo *(* scan_line)(const struct l3_xinfo_class *,
+					 char *,
+					 ino_t,
+					 enum sysfs_byteorder);
+	void * (* get_laddr)(struct l3_xinfo *);
+	void * (* get_raddr)(struct l3_xinfo *);
+	bool (* is_any_addr)(void *);
+	int family;
+	size_t size_as_cstring;
+
+};
+
+#define l2_fill_column_handler(L2, SOCK_XINFO, COLUMN_ID, STR)		\
+	do {								\
+		struct l3_xinfo_class *class = (struct l3_xinfo_class *)SOCK_XINFO->class; \
+		struct l3_xinfo *l3 = (struct l3_xinfo *)SOCK_XINFO;	\
+		void *n = NULL;						\
+		size_t ssize = class->size_as_cstring;			\
+		char *s = alloca(ssize);				\
+									\
+		switch(COLUMN_ID) {					\
+		case COL_##L2##_LADDR:					\
+			n = class->get_laddr(l3);			\
+			break;						\
+		case COL_##L2##_RADDR:					\
+			n = class->get_raddr(l3);			\
+			break;						\
+		default:						\
+			break;						\
+		}							\
+									\
+		if (n && inet_ntop(class->family, n, s, ssize)) {	\
+			*str = strdup(s);				\
+			return true;					\
```

Comment:
The macro returning from the enclosing function is iffy.

Reply:
Done. I used the "Statements and Declarations in Expressions" of gcc.